cores/arduino/stm32/analog.c

```diff
@@ -434,25 +444,114 @@ void HAL_ADC_MspInit(ADC_HandleTypeDef *hadc)
 #ifdef __HAL_RCC_ADC_CLK_ENABLE
   __HAL_RCC_ADC_CLK_ENABLE();
 #endif
+
 /* For STM32F1xx, ADC prescaler is confgured in SystemClock_Config (variant.cpp) */
 #if defined(__HAL_RCC_ADC_CONFIG) && !defined(STM32F1xx)
   /* ADC Periph interface clock configuration */
   __HAL_RCC_ADC_CONFIG(RCC_ADCCLKSOURCE_SYSCLK);
 #endif
 
-  /* Enable GPIO clock ****************************************/
-  port = set_GPIO_Port_Clock(STM_PORT(g_current_pin));
-
   /*##-2- Configure peripheral GPIO ##########################################*/
+#ifdef ADC1_BASE
+  if (hadc->Instance == ADC1)
+  {
+    port = set_GPIO_Port_Clock(STM_PORT(g_current_pinTab[0]));
+    GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pinTab[0]);
+  }
+#endif
+  
+#ifdef ADC2_BASE
+  if (hadc->Instance == ADC2)
+  {
+    port = set_GPIO_Port_Clock(STM_PORT(g_current_pinTab[1]));
+    GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pinTab[1]);
+  }
+ #endif
+  
+#ifdef ADC3_BASE
+  if (hadc->Instance == ADC3)
+  {
+    port = set_GPIO_Port_Clock(STM_PORT(g_current_pinTab[2]));
+    GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pinTab[2]);
+  }
+#endif
+  
+#ifdef ADC4_BASE
+  if (hadc->Instance == ADC4)
+  {
+    port = set_GPIO_Port_Clock(STM_PORT(g_current_pinTab[3]));
+    GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pinTab[3]);
+  }
+#endif
+
   /* ADC Channel GPIO pin configuration */
-  GPIO_InitStruct.Pin = STM_GPIO_PIN(g_current_pin);
 #ifdef GPIO_MODE_ANALOG_ADC_CONTROL
   GPIO_InitStruct.Mode = GPIO_MODE_ANALOG_ADC_CONTROL;
 #else
   GPIO_InitStruct.Mode = GPIO_MODE_ANALOG;
 #endif
   GPIO_InitStruct.Pull = GPIO_NOPULL;
   HAL_GPIO_Init(port, &GPIO_InitStruct);
+  
+  if (use_dma)
+  {
+    /*##-3- Ask DMA informations ###############################################*/
+    dma_allocate(hadc, ADC_PERIPH, &channel, &irqHandler, &indice);
+
+    /*##-4- Configure the DMA ##################################################*/
+        
+    #if defined(STM32L0xx) || defined(STM32L4xx)
+    DmaHandleTab[indice].Init.Request             = DMA_REQUEST_0;
```

Comment:
I think that indice has maybe not properly be set in this function.

Reply:
dma_allocate set the indice, so it's ok.

Reply:
that's another comment I made, dma_allocate should not set the ADC indice. This is ADC specific and should be done in this function - don't you think ?

Reply:
The indice depend on how many ADC we have, so yes it's more specific to ADC than DMA. I can move it to analog file. So prototype of dma_allocate() becomes :
void dma_allocate(void *instance, void **channel, IRQn_Type *irqHandler)

Reply:
indeed this looks better to me